### PR TITLE
	fix(common): throw an error if trackBy is not a function

### LINKED ISSUE #13388: Proposal: *ngFor - throw error when null is passed in as trackBy function
**I'm submitting a**
```
[ ] bug report => search github for a similar issue or PR before submitting
[x] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
Currently if you pass in an expression that resolves to `null` for NgFor's trackBy function it does not throw and the user does tell that user what they are doing does nothing.
For example: `*ngFor="let item of items; trackBy: item?.id"`
This syntax is from ng1 and there are examples on StackOverflow which show people using it in Angular which mislead people into thinking it is working when it is not.
This is specifically when `?` is used since if you do `trackBy: item.id` it will throw, but `trackBy: anythingElse?.something` will return null due to the safe navigation operator

**Expected behavior**
It should throw an error when `null` is passed in as the trackBy function

**Minimal reproduction of the problem with instructions**
https://plnkr.co/edit/jaY4TeiSHDYdHQpMOlVl?p=preview

**What is the motivation / use case for changing the behavior?**
People might be mislead or unknowing think that using trackBy: `item?.id` will actually track the item by that value

* **Angular version:** 2.3.0

Related #12969

### PR DESCRIPTION
**Pull request by DzmitryShylovich**
Closes https://github.com/angular/angular/issues/13388

### CODE REVIEW

File: modules/@angular/common/src/directives/ng_for.ts [hunk 1]
```diff
@@ -89,9 +89,18 @@ export class NgForRow {
 @Directive({selector: '[ngFor][ngForOf]'})
 export class NgFor implements DoCheck, OnChanges {
   @Input() ngForOf: any;
-  @Input() ngForTrackBy: TrackByFn;
+  @Input()
+  set ngForTrackBy(fn: TrackByFn) {
+    if (typeof fn !== 'function') {
```
Comment (vicb) [inline modules/@angular/common/src/directives/ng_for.ts]:
```ts
if (fn && typeof fn !== 'function') {

}
```

Would be consistent with the differ:

```ts
  constructor(private _trackByFn?: TrackByFn) {
    this._trackByFn = this._trackByFn || trackByIdentity;
  }
```

**Comment by DzmitryShylovich** [comment]
@vicb 
> `if (fn && typeof fn !== 'function') {}` Would be consistent with the differ

if I use this check instead of `if (typeof fn !== 'function')` it will be not possible to catch situation from the original [issue](https://github.com/angular/angular/issues/13388) `*ngFor="let item of items; trackBy: item?.id"`. In this case `trackBy` is null == no error

**Comment by DzmitryShylovich** [comment]
everyday http://prntscr.com/dk3j70

**Comment by vicb** [comment]
@mhevery @IgorMinar wdyt ? See the original issue also.

File: modules/@angular/common/src/directives/ng_for.ts [hunk 2]
```diff
@@ -89,9 +89,20 @@ export class NgForRow {
 @Directive({selector: '[ngFor][ngForOf]'})
 export class NgFor implements DoCheck, OnChanges {
   @Input() ngForOf: any;
-  @Input() ngForTrackBy: TrackByFn;
+  @Input()
+  set ngForTrackBy(fn: TrackByFn) {
+    console.log('input', fn);
```
Comment (pkozlowski-opensource) [inline modules/@angular/common/src/directives/ng_for.ts]: `console.log` leftover.

File: modules/@angular/common/src/directives/ng_for.ts [hunk 3]
```diff
@@ -89,9 +89,20 @@ export class NgForRow {
...
+    if (fn && typeof fn !== 'function') {
+      console.log('inside', fn);
```
Comment (pkozlowski-opensource) [inline modules/@angular/common/src/directives/ng_for.ts]: `console.log` leftover.

File: modules/@angular/common/test/directives/ng_for_spec.ts [hunk 4]
```diff
@@ -294,6 +294,19 @@ export function main() {
        }));
 
     describe('track by', () => {
+      it('should throw if trackBy is not a function', async(() => {
+           const template =
+               `<template ngFor let-item [ngForOf]="items" [ngForTrackBy]="item?.id"></template>`;
+           fixture = createTestComponent(template);
+
+        console.log('1');
```
Comment (pkozlowski-opensource) [inline modules/@angular/common/test/directives/ng_for_spec.ts]: `console.log` leftover.

File: modules/@angular/common/test/directives/ng_for_spec.ts [hunk 5]
```diff
@@ -294,6 +294,19 @@ export function main() {
...
+           getComponent().items = [{id: 1}, {id: 2}];
+        console.log('1');
```
Comment (pkozlowski-opensource) [inline modules/@angular/common/test/directives/ng_for_spec.ts]: `console.log` leftover.

File: modules/@angular/common/src/directives/ng_for.ts (hunk 2 above)
Comment (DzmitryShylovich) [inline modules/@angular/common/src/directives/ng_for.ts]: it was for debugging purposes :)

**Comment by pkozlowski-opensource** [comment]
I agree with @DzmitryShylovich that this is _very_ common mistake people do and it fails silently so having some kind of check would improve ergonomics. Having said this I've removed "LGTM" label as beside the decision this PR needs serious cleanup :-)

**Comment by DzmitryShylovich** [comment]
@pkozlowski-opensource 
> this PR needs serious cleanup

done 😄

**Comment by vicb** [comment]
We all agree that the check is needed here.

I should have given more context: The question here is should `null` be an error or not. I vote for no error as it would be a BC break and could be valid. @DzmitryShylovich thinks that `null` should be an error. Both POV are IMO acceptable but hopefully if `null` is caused by a wrong syntax, at some `null` will be defined and throw. Also IDE/linters should prevent this soon.

**Comment by mhevery** [comment]
This LGTM. I think we should throw if it is `null` 

`*ngFor="let i in is; trackBy: myFunction"` if `myFunction` is null than reverting to `identity` is wrong.

**Pull request merged**